### PR TITLE
Refactor CLI to use Click instead of our custom argparse based framework

### PR DESCRIPTION
**Pull request by thedrow**
This work is based on our decision in celery/ceps#19 to replace our custom implementation of the CLI using argparse with Click.
This change is pending the approval of celery/ceps#19 and of celery/ceps#25.

### **TODO:**
- [ ] Unit Tests
- [x] [Preload Options](https://docs.celeryproject.org/en/latest/userguide/extending.html#preload-options)
- [x] `celery worker` command
  - [x] Accept all previous arguments
  - [x] Detach main process
  - [x] Drop privileges
  - [x] Configure worker from command line
  - [x] Extract defaults from app configuration
  - [x] User options
  - [x] Documentation
- [x] `celery events` command
  - [x] Accept all previous arguments
  - [x] User options
  - [x] Documentation
- [ ] `celery beat` command
  - [x] Accept all previous arguments
  - [x] Detach main process
  - [x] Drop privileges
  - [x] Configure beat from command line
  - [ ] Extract defaults from app configuration
  - [x] User options
  - [x] Documentation
- [x] `celery shell` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery multi` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery amqp` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery status` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery inspect active` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery inspect active_queues` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery inspect clock` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery inspect conf` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery inspect memdump` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery inspect memsample` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery inspect objgraph` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery inspect ping` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery inspect query_task` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery inspect registered` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery inspect report` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery inspect reserved` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery inspect revoked` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery inspect scheduled` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery inspect stats` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery control add_consumer` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery control autoscale` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery control cancel_consumer` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery control disable_events` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery control election` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery control enable_events` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery control heartbeat` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery control pool_grow` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery control pool_restart` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery control pool_shrink` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery control rate_limit` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery control revoke` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery control shutdown` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery control terminate` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery control time_limit` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery call` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery list bindings` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery purge` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery result` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery migrate` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery graph workers` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery graph bootsteps` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery upgrade settings` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery report` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery logtool stats` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery logtool traces` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery logtool errors` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery logtool incomplete` command
  - [x] Accept all previous arguments
  - [x] Documentation
- [x] `celery logtool debug` command
  - [x] Accept all previous arguments
  - [x] Documentation

### CODE REVIEW

File: celery/bin/celery.py [hunk 1]
```diff
@@ -1,547 +1,195 @@
...
-    def load_extension_commands(self):
-        names = Extensions(self.ext_fmt.format(self=self),
-                           self.register_command).load()
-        if names:
-            command_classes.append(('Extensions', names, 'magenta'))
-
-
-if __name__ == '__main__':          # pragma: no cover
-    main()
+"""Celery Command Line Interface."""
```
Comment (auvipy) [inline celery/bin/celery.py]: nice start with click!

**Comment by dejlek** [comment]
I know I can't change the decision to use Clik - but I would suggest you to look at https://github.com/docopt/docopt first... Clik abuses annotations to the level of ridiculousness, while docopt "does it right" (TM).

**Comment by auvipy** [comment]
you are most welcome to participate in design decision. please don't hesitate to help us to choose the right tool/decision!

**Comment by thedrow** [comment]
> I know I can't change the decision to use Clik - but I would suggest you to look at https://github.com/docopt/docopt first... Clik abuses annotations to the level of ridiculousness, while docopt "does it right" (TM).

docopt is fine for basic applications.
Click provides way more features and most importantly, it supports invoking subcommands asynchronously using trio.
I was convinced by their [arguments](https://click.palletsprojects.com/en/7.x/why/#why-not-docopt-etc).

**Comment by thedrow** [comment]
This PR is now ready for initial review.
It still needs a few adjustments, a CEP and unit tests to be merged.

**Comment by jheld** [comment]
Does anyone have particular expertise in this area to help sheperd/help with extra sign off?

**Comment by thedrow** [comment]
> Does anyone have particular expertise in this area to help sheperd/help with extra sign off?

Not to my knowledge.

**Comment by georgepsarakis** [comment]
@thedrow I have some experience with Click, I think I can take a look and give some feedback.

**Comment by georgepsarakis** [comment]
@thedrow impressive work! I find the overall architecture quite intuitive and that is taking advantage of Click features. For example, [using the context](https://github.com/celery/celery/pull/5718/files#diff-af718bc9244ede32ff81adf37fc0aa7bR107) to pass through the application instance to all commands and using [option types](https://github.com/celery/celery/pull/5718/files#diff-a6e64fe5a7c24a0e587e8c92e83b87b5R49).

I didn't get the chance to check it thoroughly, since it is quite extensive, but perhaps certain commands such as `graph` or `worker` contain so much logic that perhaps could be broken down to more modules / functions.

Now that you have concluded to the final form, would it be perhaps easier for the review process, if you separated this to more branches, perhaps containing one or two commands? What do you think?

**Comment by auvipy** [comment]
https://github.com/tiangolo/typer how is that?

**Comment by thedrow** [comment]
> @thedrow impressive work! I find the overall architecture quite intuitive and that is taking advantage of Click features. For example, [using the context](https://github.com/celery/celery/pull/5718/files#diff-af718bc9244ede32ff81adf37fc0aa7bR107) to pass through the application instance to all commands and using [option types](https://github.com/celery/celery/pull/5718/files#diff-a6e64fe5a7c24a0e587e8c92e83b87b5R49).
> 
> I didn't get the chance to check it thoroughly, since it is quite extensive, but perhaps certain commands such as `graph` or `worker` contain so much logic that perhaps could be broken down to more modules / functions.
> 
> Now that you have concluded to the final form, would it be perhaps easier for the review process, if you separated this to more branches, perhaps containing one or two commands? What do you think?

Not sure how to split this work though.

**Comment by thedrow** [comment]
@georgepsarakis Ping?

**Comment by thedrow** [comment]
> https://github.com/tiangolo/typer how is that?

Typer is very nice but it is still in its early stages unfortunately.

One blocker I found is tiangolo/typer#42.
Celery uses global options for its command and while that's something that Click allows, Typer still does not.

**Comment by georgepsarakis** [comment]
@thedrow I may have some time to look at this again over the weekend.

**Comment by thedrow** [comment]
@georgepsarakis ping?

**Comment by thedrow** [comment]
@georgepsarakis Got time today?

**Comment by thedrow** [comment]
> in the meantime, would you mind rebase?

I prefer that we'll merge #6221 first.
@auvipy Do you mind taking a look?

**Comment by auvipy** [comment]
OK please allow me some time. I will have a detail look

File: celery/bin/base.py [hunk 2]
```diff
@@ -1,680 +1,229 @@
...
+            self.default = ctx.obj[self.default_value_from_context]
+        return super(CeleryOption, self).get_default(ctx)
 
-    #: Arg list used in help.
-    args = ''
+    def __init__(self, *args, **kwargs):
+        """Initialize a Celery option."""
+        self.help_group = kwargs.pop('help_group', None)
+        self.default_value_from_context = kwargs.pop('default_value_from_context', None)
+        super(CeleryOption, self).__init__(*args, **kwargs)
```
Comment (auvipy) [inline celery/bin/base.py:126]: do we need explicit Class is super() call in python 3?

File: celery/bin/base.py [hunk 3]
```diff
@@ -1,680 +1,229 @@
...
+class CeleryOption(click.Option):
+    """Customized option for Celery."""
 
-    Error = Error
-    UsageError = UsageError
-    Parser = argparse.ArgumentParser
+    def get_default(self, ctx):
+        if self.default_value_from_context:
+            self.default = ctx.obj[self.default_value_from_context]
+        return super(CeleryOption, self).get_default(ctx)
```
Comment (auvipy) [inline celery/bin/base.py:120]: same as below

File: celery/bin/base.py [hunk 4]
```diff
@@ -1,680 +1,229 @@
...
-    #: args are provided.
-    supports_args = True
+class CeleryCommand(click.Command):
+    """Customized command for Celery."""
 
-    #: List of options (without preload options).
-    option_list = None
+    def format_options(self, ctx, formatter):
+        """Write all the options into the formatter if they exist."""
+        opts = OrderedDict()
```
Comment (auvipy) [inline celery/bin/base.py:134]: as far I can recall python 3.6 made Dict Orderable?

Comment (jheld) [reply celery/bin/base.py:134]: It's an implementation detail in 3.6. Pypy happens to be that way, too. 3.7 it's official. But yes, I think the risk is low if we're 3.6+ and we don't intend to officially support any other distributions.

Comment (thedrow) [reply celery/bin/base.py:134]: I'd rather just keep it.
It may change in the future or may not. We just don't know yet.

**Comment by auvipy** [comment]
can you retarget this to master? I think we should make master as the default development branch for v5,v5.1,5.2, etc now which will be python 3.6+ only. the v4.5 branch will be used for backport works

**Comment by thedrow** [comment]
I changed the target branch to master.

**Comment by auvipy** [comment]
great

**Comment by jheld** [comment]
Congrats!!

**Comment by matllubos** [comment]
Commented code and removed tests without alternative -> merge  WTF :(

**Comment by thedrow** [comment]
We fixed the remaining issues.
We only need tests and documentation.

**Comment by mvaled** [comment]
@thedrow 

I find this left the documentation about the entry points unchanged (https://docs.celeryproject.org/en/stable/userguide/extending.html?highlight=extensions#adding-new-celery-sub-commands); but it doesn't seem to work:

```
$ celery flower
Usage: celery [OPTIONS] COMMAND [ARGS]...
Try 'celery --help' for help.

Error: No such command 'flower'.

Did you mean one of these?
    worker

$ python -c "from pkg_resources import iter_entry_points; print(list(iter_entry_points('celery.commands')))"
[EntryPoint.parse('flower = flower.command:FlowerCommand')]
```

Is this removed on purpose during this effort, or it was missed to include add the necessary glue to make the entry-points sub-commands to work?

**Comment by thedrow** [comment]
I'm working on restoring this capability with 5.0.3.

**Comment by mvaled** [comment]
@thedrow Do you need a hand or two :hand:?

**Comment by mvaled** [comment]
For the case of Flower that's gonna require a release of Flower, because they import `from celery.bin.base import Command` and that's gone.

I've achieved some slight success with:

```python
# TODO: Add all possible Flower options.
@celery.command(cls=CeleryCommand)
@click.option("--port", default=5555)
@click.option(
    "--logging",
    default="info",
    type=click.Choice(["info", "debug", "error", "warn"]),
)
@click.option(
    "--basic_auth",
    type=str,
    multiple=True,
)
@click.option(
    "--log-to-stderr",
    default=False,
    type=bool,
    is_flag=True,
)
@click.option("--url-prefix", type=str)
@click.pass_context
def flower(ctx, **kwargs):
    "Runs flower to monitor the workers"
    from tornado.options import options
    from flower.app import Flower
    from flower.urls import settings

    kwargs["basic_auth"] = list(kwargs["basic_auth"])
    for option, value in kwargs.items():
        setattr(options, option, value)
    flower = Flower(capp=ctx.obj.app, options=options, **settings)
    atexit.register(flower.stop)

    def sigterm_handler(signal, frame):
        logger.info("SIGTERM detected, shutting down")
        sys.exit(0)

    signal.signal(signal.SIGTERM, sigterm_handler)

    try:
        flower.start()
    except (KeyboardInterrupt, SystemExit):
        pass


celery.add_command(flower)
```

But the `url-prefix` option is not working properly.

**Comment by mvaled** [comment]
Well now I know that several `options` in flower are "massaged" and converted to settings.  For the case of `url-prefix` the do this before calling `Flower`:

```python
    if options.url_prefix:
        for name in ["login_url", "static_url_prefix"]:
            settings[name] = prepend_url(settings[name], options.url_prefix)
```

**Comment by thedrow** [comment]
Let's discuss this over a video chat.
Drop me an email?

**Pull request merged**